### PR TITLE
Fix rendering of geometry generator expressions which return geometry collection results

### LINKED ISSUE #35356: Intersection (Geom. Generator) not rendering when polygons intersect in one part and abut in another
This bug is a little bit difficult to explain, but it took me the better part of a full day to figure out this was at issue in one of my own projects.

The bug seems to happen when using Geometry Generator: Intersection...when a portion of polygons do in fact intersect (A), then have a void area (B), then have a portion that abuts (C).  The actual intersect area does not render.

Screenshot from 2020-03-25 16-41-11

By moving any point from the abut portion (D), the intersect area does render (E).

Screenshot from 2020-03-25 16-51-29

**How to Reproduce**

1. Load provided testA.gpkg
2. Apply testA.qml for styles, including intersection code:

`intersection(
	geometry(
		get_feature('testA', 'id', '1')
	),
	geometry(
		get_feature('testA', 'id', '2')
	)
)`

3. Geometry Generator does NOT render intersect area.
4. Now, edit any points from abut portion and Geometry Generator does render intersect area(s).

**QGIS and OS versions**

QGIS version | 3.13.0-Master | QGIS code revision | 5a3b1ced84
-- | -- | -- | --
Compiled against Qt | 5.9.5 | Running against Qt | 5.9.5
Compiled against GDAL/OGR | 2.2.3 | Running against GDAL/OGR | 2.2.3
Compiled against GEOS | 3.6.2-CAPI-1.10.2 | Running against GEOS | 3.6.2-CAPI-1.10.2 4d2925d6
Compiled against SQLite | 3.22.0 | Running against SQLite | 3.22.0
PostgreSQL Client Version | 10.12 (Ubuntu 10.12-0ubuntu0.18.04.1) | SpatiaLite Version | 4.3.0a
QWT Version | 6.1.3 | QScintilla2 Version | 2.10.2
PROJ.4 Version | 493
OS Version | Linux Mint 19.3 | This copy of QGIS writes debugging output.
Active python plugins | latlontools; joinmultiplelines; processing; db_manager

[testA.zip](https://github.com/qgis/QGIS/files/4383771/testA.zip)

### PR DESCRIPTION
**Pull request by nyalldawson**
Fixes #35356

But in general, this is exciting because it allows QgsSymbol to correctly render any
input geometry types, and handle collections nicely. In future releases we could quite
easily add a symbol layer type for rendering collections, where the symbol layer has
a marker/line/fill symbol set, and the appropriate selection of these is used to
render each geometry part (it's basically identical to a geometry generator!)

Geometry collection handling at a QgsVectorLayer level, now that's a totally different beast....

### CODE REVIEW

**Comment by nyalldawson** [comment]
Also note for reviewers: I think QgsSymbol::renderFeature is overdue for splitting up into smaller functions, but I was conscious of avoiding too much code churn in this commit. So that can come later...

**Comment by nyalldawson** [comment]
Another note for reviewers:

What has changed here (the diff is more confusing then it needs to be) is:

1. instead of basing all decisions off the geometry type, we now base them off the symbol type. Previously, the different parts of this function would test the geometry type to select between handling points, lines or polygons. Now we use the actual symbol type to do this.
2. During the first stage, we dive into collections/multipart geometries with a recursive function which filters and collects only the geometry parts which are compatible with the symbol type. 
3. The remaining stages all use the filtered part lists to render and postprocess geometry parts, instead of reusing the original geometry. 

So basically, if the function gets a mixed collection geometry (as it was with the geometry generator result), the first pass collects all the parts of this collection which match the symbol type, and then we only consider these in the rendering stage.

**Review by nirvn — APPROVED** [review]
